src/compile/data/optimize.ts

```diff
@@ -114,6 +115,33 @@ function getLeaves(roots: DataFlowNode[]) {
   return leaves;
 }
 
+export function isParseMergeable(x: ParseNode, y: ParseNode) {
+  const xParse = x.parse;
+  const yParse = y.parse;
+  for (const i of keys(xParse)) {
+    if (i in yParse && yParse[i] !== xParse[i]) {
+      return false;
+    }
+  }
+  return true;
+}
+
+export function mergeParse(node: DataFlowNode) {
+  const parseChildren = node.children.filter((x): x is ParseNode => x instanceof ParseNode);
```

Comment:
what is this is syntax?

`.filter((x:ParseNode) => x instanceof ParseNode);`

Reply:
That's the return `type`. Equivalent to

Reply:
Unless the _return_ type is coerced, type narrowing doesn't happen.

Reply:
Ohh, I see. Makes sense.